### PR TITLE
Implement media capture streams.

### PR DESCRIPTION
**Pull request by jdm**
This gets audio capture working in the example program for me.

### CODE REVIEW

**Comment by jdm** [comment]
I grabbed all of https://github.com/Manishearth/media-capture/blob/master/src/main.rs and dumped it into the gstreamer backend, then pruned anything that I didn't actually need for this particular change.

**Comment by jdm** [comment]
@manishearth see the second commit.

File: backends/gstreamer/src/media_stream.rs [hunk 1]
```diff
@@ -53,18 +54,23 @@ impl GStreamerMediaStream {
             StreamType::Audio => &*RTP_CAPS_OPUS,
             StreamType::Video => &*RTP_CAPS_VP8,
         };
-        println!("atttaching a {:?} stream", self.type_);
         self.elements.last().as_ref().unwrap().link_filtered(webrtcbin, caps).unwrap();
+        pipeline.set_state(gst::State::Playing).into_result().unwrap();
```
Comment (Manishearth) [inline backends/gstreamer/src/media_stream.rs]: is this necessary? the program crashes (with the state change unwrap in prepare_for_negotiation) unless i remove this line

**Comment by Manishearth** [comment]
other than the minor fact that this doesn't seem to work, these changes look good

File: backends/gstreamer/src/media_stream.rs (hunk 1 above)
Comment (jdm) [reply backends/gstreamer/src/media_stream.rs]: Whereas if I do not include this line, it does not work at all for me ;)

Comment (Manishearth) [reply backends/gstreamer/src/media_stream.rs]: So when I remove this line it throws up one blank video screen, i.e. it still doesn't work, but it doesn't panic either.

Comment (Manishearth) [reply backends/gstreamer/src/media_stream.rs]: Testing the patch against the JS example has it connect (without panicking) and turn on my camera, but the JS side doesn't show anything and the testsink doesn't show up either.

The panic happens when I try to get this to talk to itself, and it's the offer side that fails.

**Comment by jdm** [comment]
Ok, the newest commit allows it to continue working for me without directly changing the playback state of the pipeline.

**Comment by Manishearth** [comment]
Works now, I get fun echoes and the video works in one direction.

The JS version now throws an error on the JS page, `Status: OperationError: Cannot set ICE candidate for level=1 mid=: No such transceiver.`


(it's possible that having two programs attempting to acquire camera access also doesn't work??)

**Comment by Manishearth** [comment]
got it working! The issue is indeed that two programs are attempting to acquire the camera at once, which my laptop doesn't like.
 


For the case where it talks to itself, I made it work by having only one side use the actual input sources.

Still doesn't work on the js example for me, presumably because of the same camera issue. But that's fine, I primarily test this with itself, and on your machine this isn't an issue.


```patch
diff --git a/examples/simple_webrtc.rs b/examples/simple_webrtc.rs
index 3dcbfde..afe5916 100644
--- a/examples/simple_webrtc.rs
+++ b/examples/simple_webrtc.rs
@@ -139,14 +139,18 @@ impl State {
         self.signaller = Some(s);
         let webrtc = self.webrtc.as_ref().unwrap();
         webrtc.init();
-        let video = self
+        let (video, audio) = if self.peer_id.is_some() {
+            (self
             .media
             .create_videoinput_stream()
-            .unwrap_or_else(|| self.media.create_videostream());
-        let audio = self
+            .unwrap_or_else(|| self.media.create_videostream()),
+            self
             .media
             .create_audioinput_stream()
-            .unwrap_or_else(|| self.media.create_audiostream());
+            .unwrap_or_else(|| self.media.create_audiostream()))
+        } else {
+            (self.media.create_videostream(), self.media.create_audiostream())
+        };
         webrtc.add_stream(&*video);
         self.streams.push(video);
         webrtc.add_stream(&*audio);
```

**Pull request merged**